sql-plugin/src/main/scala/org/apache/spark/sql/rapids/mathExpressions.scala

```diff
@@ -453,13 +453,160 @@ abstract class GpuRoundBase(child: Expression, scale: Expression) extends GpuBin
+    //  the max value of Short: 32767
+    //  The first digit is up to 3, which can't be rounded up.
+    //  the max value of Int32: 2147483647
+    //  The first digit is up to 2, which can't be rounded up.
+    //  the max value of Float32: 3.4028235E38
+    //  The first digit is up to 3, which can't be rounded up.
+    //  the max value of Float64: 1.7976931348623157E308
+    //  The first digit is up to 1, which can't be rounded up.
+    //  the max value of Int64: 9223372036854775807
+    //  The first digit is up to 9, which can be rounded up.
+    //
+    // When rounding up 19-digits long values on the first digit, the result can be 1e19 or -1e19.
+    // Since LongType can not hold these two values, the 1e19 overflows as -8446744073709551616L,
+    // and the -1e19 overflows as 8446744073709551616L. The overflow happens in the same way for
+    // HALF_UP (round) and HALF_EVEN (bround).
+    def fixUpInt64OnBounds(zeroFn: () => Scalar): ColumnVector = {
+      // Builds predicates on whether there is a round up on the max digit or not
+      val litForCmp = Seq(Scalar.fromLong(1000000000000000000L),
+                          Scalar.fromLong(4L),
+                          Scalar.fromLong(-4L))
+      val (needRep, needNegRep) = withResource(litForCmp) { case Seq(base, four, minusFour) =>
+        withResource(lhsValue.div(base)) { headDigit =>
+          closeOnExcept(headDigit.greaterThan(four)) { posRep =>
+            closeOnExcept(headDigit.lessThan(minusFour)) { negRep =>
+              posRep -> negRep
+            }
+          }
+        }
+      }
+      // Replaces with corresponding literals
+      val litForRep = Seq(zeroFn(),
+                          Scalar.fromLong(8446744073709551616L),
+                          Scalar.fromLong(-8446744073709551616L))
+      val repVal = withResource(litForRep) { case Seq(zero, upLit, negUpLit) =>
+        withResource(needRep) { _ =>
+          withResource(needNegRep) { _ =>
+            withResource(needNegRep.ifElse(upLit, zero)) { negBranch =>
+              needRep.ifElse(negUpLit, negBranch)
+            }
+          }
+        }
+      }
+      // Handles null values
+      withResource(repVal) { _ =>
+        if (lhsValue.hasNulls) {
+          repVal.mergeAndSetValidity(BinaryOp.BITWISE_AND, lhsValue)
+        } else {
+          repVal.incRefCount()
+        }
+      }
+    }
+
+    // Fixes up float points rounded by a scale exceeding the max digits of data type. Under this
+    // circumstance, cuDF produces different results to Spark.
+    // Compared to integral values, fixing up round overflow of float points needs to take care
+    // of some special values: nan, inf, -inf.
+    def fpZeroReplacement(zeroFn: () => Scalar,
+        infFn: () => Scalar,
+        negInfFn: () => Scalar): ColumnVector = {
+      val scaleVal = scale.getValue.asInstanceOf[Int]
+      val maxDigits = if (dataType == FloatType) 39 else 309
```

Comment:
I prefer to have a similar API to `DecimalUtil.getPrecisionForIntegralType`, e.g. `DecimalUtil.getPrecisionForFloatType`

Reply:
Strictly speaking, the `39` and `309` are numeric bounds rather than precision of float types. These numeric bounds are rarely used in spark-rapids.  And they have little relation with `DecimalUtil`. So, I don't think we need to make it a common method.